Kerberos.NET/Entities/Pac/PrivilegedAttributeCertificate.cs

```diff
@@ -289,57 +306,95 @@ public ReadOnlyMemory<byte> Encode(KerberosKey kdcKey, KerberosKey serverKey)
 
         private static void SignPac(IEnumerable<PacObject> pacElements, Memory<byte> pacUnsigned, KerberosKey kdcKey, KerberosKey serverKey)
         {
-            foreach (var element in pacElements.Where(e => e is PacSignature).Cast<PacSignature>())
+            PacSignature serverSignature = null;
+
+            foreach (var element in pacElements.Where(e => e is PacSignature).OrderBy(e => e.PacType).Cast<PacSignature>())
             {
                 if (element.PacType == PacType.SERVER_CHECKSUM)
                 {
+                    serverSignature = element;
+
                     element.Sign(pacUnsigned, serverKey);
                 }
 
                 if (element.PacType == PacType.PRIVILEGE_SERVER_CHECKSUM)
                 {
-                    element.Sign(pacUnsigned, kdcKey);
+                    element.Sign(serverSignature.Signature, kdcKey);
                 }
             }
         }
 
-        private static Memory<byte> GeneratePac(IEnumerable<PacObject> pacElements)
+        private static int Align(int position, int mask)
         {
-            var buffer = new NdrBuffer(align: false);
+            var shift = position & mask - 1;
 
-            buffer.WriteInt32LittleEndian(pacElements.Count());
-            buffer.WriteInt32LittleEndian(PAC_VERSION);
+            if (shift != 0)
+            {
+                var seek = 8 - shift;
+
+                return seek;
+            }
+
+            return 0;
+        }
+
+        [DebuggerDisplay("{Type} {Offset} {Length}")]
+        private struct PacBuffer
+        {
+            public PacType Type { get; set; }
+
+            public int Length { get; set; }
+
+            public int Offset { get; set; }
+
+            public PacObject Element { get; set; }
+        }
+
+        private static Memory<byte> GeneratePac(IEnumerable<PacObject> pacElements)
+        {
+            var offset = 8 + (pacElements.Count() * 16);
 
-            var headerLength = 8 + (pacElements.Count() * 16);
-            var offset = headerLength;
+            var buffers = new PacBuffer[pacElements.Count()];
```

Comment:
I would evaluate `pacElements.Count()` just once.
So 
```c#
var pacCount = pacElements.Count();
var offset = 8 + (pacCount * 16);
var buffers = new PacBuffer[pacCount];
```

Depending on the actual type of the `IEnumerable<PacObject>` it may or may not be a difference on getting the `Count()` multiple times. But in general, where the actual type isn't known, it may harm, as in the worst case the enumerable is iterated twice which may have side-effects.

Here the type is `List<PacObject>`, so this shouldn't be a problem, but good practice :wink:

--- 

Or:

I would make the argument `List<PacObject>` (internal usage only), so one can use `Count` and indexed access in the below for-loop.

Reply:
Fixed.